# SQL Project

### Description of the project

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers.
You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

### Description of the data

#### books: Contains data on books:

book_id

author_id

title

num_pages — number of pages

publication_date

publisher_id

#### authors: Contains data on authors:

author_id

author

#### publishers: Contains data on publishers:

publisher_id


publisher

#### ratings: Contains data on user ratings:

rating_id

book_id

username — the name of the user who rated the book

rating

#### reviews: Contains data on customer reviews:

review_id

book_id

username — the name of the user who reviewed the book

text — the text of the review

### Task

1. Find the number of books released after January 1, 2000.

2. Find the number of user reviews and the average rating for each book.

3. Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude 
brochures and similar publications from your analysis).

4. Identify the author with the highest average book rating: look only at books with at least 50 ratings.

5. Find the average number of text reviews among users who rated more than 50 books.

### importing libraries

In [1]:
# import libraries
import pandas as pd
!pip install psycopg2-binary

from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,        # connection port
             'db': 'data-analyst-final-project-db'}   #the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Checking the data

In [2]:
def sql(query):
    return pd.io.sql.read_sql(query, con = engine)

In [3]:
a='''SELECT * 
    FROM books;'''

sql(a)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
b='''SELECT * 
    FROM authors;'''

sql(b)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
c='''SELECT * 
    FROM publishers;'''

sql(c)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [6]:
d='''SELECT * 
    FROM ratings;'''

sql(d)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [7]:
e = '''SELECT * 
    FROM reviews;'''

sql(e)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Conclusions: The databases contain information on 1000 books, 636 authors, 340 publishers, 6456 ratings and 2793 reviews - all in separate tables. Most of the columns have book id in common with the publisher and author tables only having two values, forming a dictionary to their ID's.

In [8]:
# the reviewer cell: 
pd.io.sql.read_sql(
    ''' SELECT *
    FROM books
    LIMIT 1
    '''
, con = engine)                    

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

In [9]:
# the reviewer cell: also, we could estimate the table size within loading the first line of the table
# for that, we will add the `table_size` column using simple window function:

pd.io.sql.read_sql('''
SELECT *,
       COUNT(*) OVER() AS table_size
FROM books 
LIMIT 1        
'''
, con = engine)

book_id  author_id         title  num_pages publication_date  publisher_id  \
0        1        546  'Salem's Lot        594       2005-11-01            93   

   table_size  
0        1000

### Finding the number of books released after January 1, 2000

In [10]:
f = '''SELECT COUNT(*) AS cnt 
    FROM books
    WHERE publication_date > '2000-01-01';'''

sql(f)

cnt
0  819

Conclusions: Out of 1000 entries 819 of those in the database were released after January 1, 2000.

### Finding the number of user reviews and average rating for each book

In [11]:
g= '''SELECT COUNT(DISTINCT review_id) AS reviews_count, AVG(rating) AS avg_rating, books.book_id
FROM(( books
LEFT JOIN ratings ON ratings.book_id = books.book_id)
LEFT JOIN reviews ON reviews.book_id = ratings.book_id)
GROUP BY books.book_id
ORDER BY book_id ASC; '''

sql(g)

reviews_count  avg_rating  book_id
0                2    3.666667        1
1                1    2.500000        2
2                3    4.666667        3
3                2    4.500000        4
4                4    4.000000        5
..             ...         ...      ...
995              3    3.666667      996
996              3    3.400000      997
997              4    3.200000      998
998              2    4.500000      999
999              4    3.833333     1000

[1000 rows x 3 columns]

Conclusions: Review count seems to vary widely and the rating system is on a scale to 5.

In [12]:
# the reviewer's code:
g= '''
SELECT COUNT(*)
FROM(
SELECT COUNT(DISTINCT review_id) AS reviews_count, AVG(rating) AS avg_rating, books.book_id
FROM(( books
INNER JOIN ratings ON ratings.book_id = books.book_id)
INNER JOIN reviews ON reviews.book_id = ratings.book_id)
GROUP BY books.book_id
) AS main_query

'''

sql(g)

count
0    994

In [13]:
# the reviewer's code:
g= '''
SELECT COUNT(*)
FROM books
'''
sql(g)

count
0   1000

### Identify the publisher that has released the greatest number of books

In [14]:
h = '''SELECT books.publisher_id AS id,
    publishers.publisher AS publisher,
    COUNT(books.book_id) AS book_count
    
    FROM books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
    
    WHERE books.num_pages > 50
    
    GROUP BY books.publisher_id,
    publishers.publisher
    
    ORDER BY book_count DESC
    
    LIMIT 1;'''

sql(h)

id      publisher  book_count
0  212  Penguin Books          42

Conclusions:
The publisher from the databases that released the most books was Penguin. They're the number 1 or 2 publisher in the world - fun fact, I used to work there.

### Identify the author with the highest average book rating

In [15]:
highest_avg_rating_author_query = '''
SELECT a.author, 
    a.author_id, 
    AVG(r.rating) as avg_rating 
FROM authors 
AS a INNER JOIN books AS k ON a.author_id = k.author_id 
INNER JOIN ratings AS r ON r.book_id=k.book_id 
WHERE k.book_id IN 
    (SELECT b.book_id 
    FROM books 
    AS b INNER JOIN ratings AS s ON b.book_id = s.book_id 
    GROUP BY b.book_id HAVING COUNT(DISTINCT s.rating_id)>50) 
GROUP BY a.author_id 
ORDER BY avg_rating DESC LIMIT 1;
'''

highest_avg_rating_author_query = pd.io.sql.read_sql(highest_avg_rating_author_query, con = engine)
highest_avg_rating_author_query

author  author_id  avg_rating
0  J.K. Rowling/Mary GrandPré        236    4.287097

Conclusions:
Author J.K. Rowling/Mary GrandPré has the highest average book rating among the books having at least 50 (more than 49) ratings.

In [16]:
# the reviewer's cell
query = '''
SELECT a.author, 
       a.author_id, 
       AVG(b.avg_rating) as avg_rating, 
       COUNT(b.avg_rating) as book_count 
FROM authors AS a 
INNER JOIN (SELECT b.book_id, b.author_id, AVG(r.rating) as avg_rating 
            FROM books AS b 
            INNER JOIN ratings AS r ON b.book_id = r.book_id 
            GROUP BY b.book_id, b.author_id
            HAVING COUNT(DISTINCT r.rating_id)>=50) AS b ON a.author_id = b.author_id 
GROUP BY a.author_id 
ORDER BY avg_rating DESC 
'''

pd.io.sql.read_sql(query, con = engine)


author  author_id  avg_rating  \
0                          J.K. Rowling/Mary GrandPré        236    4.283844   
1                   Markus Zusak/Cao Xuân Việt Khương        402    4.264151   
2                                      J.R.R. Tolkien        240    4.258446   
3                                   Louisa May Alcott        376    4.192308   
4                                        Rick Riordan        498    4.080645   
5                                     William Golding        621    3.901408   
6                                       J.D. Salinger        235    3.825581   
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce        469    3.789474   
8   William Shakespeare/Paul Werstine/Barbara A. M...        630    3.787879   
9                                           Dan Brown        106    3.754540   
10                                         Lois Lowry        372    3.750000   
11            George Orwell/Boris Grabnar/Peter Škerl        195    3.729730   
12                                    Stephenie Meyer        554    3.662500   
13                                     John Steinbeck        311    3.622951   

    book_count  
0            4  
1            1  
2            2  
3            1  
4            1  
5            1  
6            1  
7            1  
8            1  
9            2  
10           1  
11           1  
12           1  
13           1

### Find the average number of text reviews among users who rated more than 50 books

In [17]:
j = '''SELECT AVG(SUBQ.reviews_per_user) AS avg_reviews
    
    FROM 
        (SELECT reviews.username AS username,
        COUNT(DISTINCT reviews.review_id) AS reviews_per_user
    
        FROM reviews
        LEFT JOIN ratings ON ratings.username = reviews.username
    
        GROUP BY reviews.username) AS SUBQ
        
    WHERE username in 
        (SELECT username as username
    
        FROM ratings
        
        GROUP BY username
       
        HAVING COUNT(rating_id) > 50);'''

sql(j)

avg_reviews
0    24.333333

Conclusions:
The average number of text reviews for users, who had rated more than 50 books, is 24.3.

### Final Conclusions

Out of 1000 books, their associated authors, publishers, user reviews and ratings, we've learned that more than 80% of them were published afer Jan 1 2001, and the most prolific publisher was Penguin. The highest avg. rating author from the data was J.K. Rowling and users who left at least 50 ratings had written on average 24.3 reviews. What does this equate too? The publishing world, or at least this dataset, is dominated by big publishers (Penguin), household names (Rowling) and some devoted fans. Arguably any product developed would need to keep an eye on the integration of these powerhouses and the users who love them.